doc/specs/#5000 - Process Model 2.0/#5000 - Process Model 2.0.md

```diff
@@ -0,0 +1,882 @@
+
+We should reserve the session id `0` to always refer to "The current window", if
+there is one. So `wt -s 0 new-tab` will run `new-tab` in the current window (if
+we're being run from WT).
+
+If `wt -s 0 <commands>` is run _outside_ a WT instance, it could attempt to glom
+onto _the most recent WT window_ instead. This seems more logical than something
+like `wt --session last` or some other special value indicating "run this in the
+MRU window".
+
+That might be a simple, but **wrong**, implementation for "the current window".
+If the servants always raise an event when their window is focused, and the
+monarch keeps track of the MRU order for servants, then one could naively assume
+that the execution of `wt -s 0 <commands>` would always return the window the
+user was typing in, the current one. However, if someone were to do something
+like `sleep 10 ; wt -s 0 <commands>`, then the user could easily focus another
+WT window during the sleep, which would cause the MRU window to not be the same
+as the window executing the command.
+
+I'm not sure that there is a better solution for the `-s 0` scenario other than
+attempting to use the `WT_SESSION` environment variable. If a `wt.exe` process is
+spawned and that's in it's environment variables, it could try and ask the
+monarch for the servant who's hosting the session corresponding to that GUID.
+This is more of a theoretical solution than anything else.
+
+In Single-Instance mode, running `wt -s 0` outside a WT window will still cause
+the commandline to glom to the existing single terminal instance, if there is
+one.
+
+#### Scenario: Quake Mode
+
+"Quake mode" has a variety of different scenarios<sup>[[1]](#footnote-1)</sup>
+that have all been requested, more than what fits cleanly into this spec.
+However, there's one key aspect of quake mode that is specifically relevant and
+worth mentioning here.
+
+One of the biggest challenges with registering a global shortcut handler is
+_what happens when multiple windows try to register for the same shortcut at the
+same time_? From my initial research, it seems that only the first process to
+register for the shortcut will succeed. This makes it hard for multiple windows
+to handle the global shortcut key gracefully. If a second window is created, and
+it fails to register the global hotkey, then the first window is closed, there's
+no way for the second process to track that and re-register as the handler for
+that key.
+
+With the addition of monarch/servant processes, this problem becomes much easier
+to solve. Now, the monarch process will _always_ be the process to register the
+shortcut key, whenever it's elected. If it dies and another servant is elected
+monarch, then the new monarch will register as the global hotkey handler.
+
+Then, the monarch can use it's pre-established channels of communication with
+the other window processes to actually drive the response we're looking for.
+

```

Comment:
I worry about this being another responsibility of the monarch process. What if there are no wt.exes running on the system already? Do we start a window process monarch with no children that hides its own window to take the responsibility of this keyboard handler immediately after install?

I would have almost thought that the keyboard portion of Quake Mode could be handled by a long running service inside the Terminal package that can be waiting for this event and either send the message to the monarch or start it as appropriate (with the arguments if necessary to look like a slide-in drop down from the top).

Reply:
I presumed that quake mode wouldn't be enabled until the Terminal is started the first time, so I guess a lot of this comes from that assumption. I suppose that a long-running service could _also_ work for this case. It'd still be the monarch job to communicate the activation appropriately to the correct servant, but I'll note this as a possibility.